util/logutil/inconsist/reporter.go

```diff
@@ -0,0 +1,264 @@
+// limitations under the License.
+
+package inconsist
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/errno"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/helper"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/dbterror"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/tikv/client-go/v2/tikv"
+	"go.uber.org/zap"
+)
+
+var (
+	// ErrAdminCheckInconsistent returns for data inconsistency for admin check.
+	ErrAdminCheckInconsistent = dbterror.ClassAdmin.NewStd(errno.ErrDataInConsistent)
+	// ErrLookupInconsistent returns for data inconsistency for index lookup.
+	ErrLookupInconsistent = dbterror.ClassExecutor.NewStd(errno.ErrDataInconsistentMismatchCount)
+	// ErrAdminCheckInconsistentWithColInfo returns for data inconsistency for admin check but with column info.
+	ErrAdminCheckInconsistentWithColInfo = dbterror.ClassExecutor.NewStd(errno.ErrDataInconsistentMismatchIndex)
+)
+
+func getMvccByKey(sctx sessionctx.Context, key kv.Key, decodeMvccFn func(kv.Key, *kvrpcpb.MvccGetByKeyResponse, map[string]interface{})) string {
+	if key == nil {
+		return ""
+	}
+	tikvStore, ok := sctx.GetStore().(helper.Storage)
+	if !ok {
+		return ""
+	}
+	h := helper.NewHelper(tikvStore)
+	data, err := h.GetMvccByEncodedKey(key)
+	if err != nil {
+		return ""
+	}
+	regionID := getRegionIDByKey(tikvStore, key)
+
+	decodeKey := strings.ToUpper(hex.EncodeToString(key))
+
+	resp := map[string]interface{}{
+		"key":      decodeKey,
+		"regionID": regionID,
+		"mvcc":     data,
+	}
+
+	if decodeMvccFn != nil {
+		decodeMvccFn(key, data, resp)
+	}
+
+	rj, err := json.Marshal(resp)
+	if err != nil {
+		return ""
+	}
+	return string(rj)
+}
+
+func getRegionIDByKey(tikvStore helper.Storage, encodedKey []byte) uint64 {
+	keyLocation, err := tikvStore.GetRegionCache().LocateKey(tikv.NewBackofferWithVars(context.Background(), 500, nil), encodedKey)
+	if err != nil {
+		return 0
+	}
+	return keyLocation.Region.GetID()
+}
+
+// Reporter is a helper to generate report.
+type Reporter struct {
+	HandleEncode func(handle kv.Handle) kv.Key
+	IndexEncode  func(idxRow *RecordData) kv.Key
+	Tbl          *model.TableInfo
+	Idx          *model.IndexInfo
+	Sctx         sessionctx.Context
+}
+
+func (r *Reporter) decodeRowMvccData(_ kv.Key, respValue *kvrpcpb.MvccGetByKeyResponse, outMap map[string]interface{}) {
+	colMap := make(map[int64]*types.FieldType, 3)
+	for _, col := range r.Tbl.Columns {
+		colMap[col.ID] = &col.FieldType
+	}
+
+	if respValue.Info != nil {
+		var err error
+		datas := make(map[string][]map[string]string)
+		for _, w := range respValue.Info.Writes {
+			if len(w.ShortValue) > 0 {
+				datas[strconv.FormatUint(w.StartTs, 10)], err = decodeMvccData(w.ShortValue, colMap, r.Tbl)
```

Comment:
May need to put the error information into resullt if it's not nil?

Reply:
it follows previous mvcc  api decode logic, maybe its useful to let user know the reason why data doesn't decode~? but remove it is ok for me too